File: Python/AllPathsfromSourceLeadtoDestination.py
from typing import List
from typing import Dict

class Solution:

    def leadsToDestination(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        graph = {}
        self.res = True
        for e in edges:
            if e[0] not in graph:
                graph[e[0]] = []
            graph[e[0]].append(e[1])

        visited = [False] * n
        stack = [False] * n
        self.dfs(graph, source, visited, stack, destination)

        return self.res

    def dfs(self, graph: Dict, n: int, visited: List[bool], stack: List[bool], dst: int):
        if not self.res:
            return
        
        stack[n] = True
        visited[n] = True

        if n not in graph:
            if n == dst:
                stack[n] = False
                return
            else:
                self.res = False
                return

        for adj in graph[n]:
            if stack[adj]:
                self.res = False
                return
            if not visited[adj]:    
                self.dfs(graph, adj, visited, stack, dst)

        stack[n] = False

test = Solution()
edges = [[0,1],[0,2],[1,3],[2,3]]
print(test.leadsToDestination(4, edges, 0, 3))

File: Java/src/AllPathsfromSourceLeadtoDestination.java
import java.util.HashSet;

public class AllPathsfromSourceLeadtoDestination {

    private boolean res;
    public boolean leadsToDestination(int n, int[][] edges, int source, int destination) {
        HashSet<Integer>[] graph = new HashSet[n];
        for(int i = 0; i < n; i++) graph[i] = new HashSet<>();
        for(int[] e : edges) graph[e[0]].add(e[1]);
        res = true;
        boolean[] visited = new boolean[n];
        boolean[] onStack = new boolean[n];

        dfs(source, graph, visited, onStack, destination);

        return res;
    }

    private void dfs(int n, HashSet<Integer>[] graph, boolean[] visited, boolean[] onStack, int dst){
        if(!res) return;
        visited[n] = true;
        onStack[n] = true;

        if(graph[n].size() == 0 && dst != n) {
            res = false;
            return;
        }

        for(int adj : graph[n]){
            if(onStack[adj]) {
                res = false;
                return;
            }else{
                if(!visited[adj]) dfs(adj, graph, visited, onStack, dst);
            }
        }

        onStack[n] = false;
    }

    public static void main(String[] args) {
        AllPathsfromSourceLeadtoDestination test = new AllPathsfromSourceLeadtoDestination();
        int[][] edges = {{0,1},{1,1},{1,2}};
        System.out.println(test.leadsToDestination(4, edges, 0, 2));
    }


}